---
title: "Article APO"
output: github_document
---

# Pour les archées

### Définition de la variable et affichage des fichiers

```{r}
path <- "~/article-apo/Fichierdesequences"
list.files(path)
data<-read.table(file = "/home/rstudio/article-apo/SraRunTable.txt", header = TRUE, sep=",") # Création de la table de données
```

### Utiliser une colonne des métadonnées pour définir le nom des échantillons

```{r}
fnFs <- sort(list.files(path, pattern="_1.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_2.fastq.gz", full.names = TRUE))
print(fnFs)
print(fnRs)
```

```{r}
sample.names <- (data$Sample.Name)
sample.names
```

```{r}
plotQualityProfile(fnFs[1:20])
```

```{r}
plotQualityProfile(fnRs[1:20])
```

### Filtrer et rogner

```{r}
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names
```

```{r}
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(260,260),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) 
head(out)
```

### Construire un modèle de probabilité d’erreur

```{r}
errF <- learnErrors(filtFs, multithread=TRUE)
```

```{r}
errR <- learnErrors(filtRs, multithread=TRUE)
```

```{r}
plotErrors(errF, nominalQ=TRUE)
```

### Application de l’algorithme d’interférence d’échantillon de base aux données de séqences filtrées et découpées

```{r}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
```

```{r}
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)
```

```{r}
dadaFs[[1]]
```

### Fusionner les lectures appariées

```{r}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)
```

```{r}
plotQualityProfile(filtFs[1:10])
```

```{r}
plotQualityProfile(filtRs[1:10])
```

```{r}
head(mergers[[1]])
```

### Construire une table de séquences

```{r}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)
```

```{r}
table(nchar(getSequences(seqtab)))
```

### Supprimer les chimères

```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
```

```{r}
dim(seqtab.nochim)
```

```{r}
sum(seqtab.nochim)/sum(seqtab)
```

### Suivre les lectures à travers le pipeline

```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
```

### Attribuer une taxonomie 

```{r}
taxa <- assignTaxonomy(seqtab.nochim, "~/article-apo/silva_nr_v132_train_set.fa.gz", multithread=TRUE)
```

```{r}
taxa.print <- taxa 
rownames(taxa.print) <- NULL
head(taxa.print)
```

### Evaluer la précision

```{r}
# Vérifier les noms des lignes de seqtab.nochim
rownames(seqtab.nochim)

# Vérifier la structure complète de l'objet
str(seqtab.nochim)

# Exemple de recherche du nom exact
grep("Mock", rownames(seqtab.nochim))
```

```{r}
unqs.mock <- seqtab.nochim
unqs.mock <- sort(unqs.mock[unqs.mock>0], decreasing=TRUE) # Drop ASVs absent in the Mock
cat("DADA2 inferred", length(unqs.mock), "sample sequences present in the Mock community.\n")
```

```{r}
mock.ref <- getSequences(file.path(path, "HMP_MOCK.v35.fasta"))
match.ref <- sum(sapply(names(unqs.mock), function(x) any(grepl(x, mock.ref))))
cat("Of those,", sum(match.ref), "were exact matches to the expected reference sequences.\n")
```

```{r}
# Vérifier les types de unqs.mock et mock.ref
class(unqs.mock)
class(mock.ref)

# Vérifier la structure des objets
str(unqs.mock)
str(mock.ref)
```

```{r}
unqs.mock <- unlist(unqs.mock)
```

```{r}
unqs.mock[[1]]  # Accède au premier élément de la liste
```

```{r}
# Assurez-vous que unqs.mock et mock.ref sont des vecteurs de chaînes
unqs.mock <- unlist(unqs.mock)
mock.ref <- unlist(mock.ref)

# Appliquer la logique avec sapply et grepl
result <- sum(sapply(unqs.mock, function(x) any(grepl(x, mock.ref))))

# Vérifier le résultat
print(result)
```

```{r}
# Vérifier si le fichier existe à l'emplacement spécifié
file_exists <- file.exists(mock.ref)
print(file_exists)
```

```{r}
mock.ref <- "/home/rstudio/ARTICLE-APO/FICHIERDESEQUENCES/HMP_MOCK.V35.FASTA"
# Exemple pour donner des permissions de lecture
chmod +r /home/rstudio/ARTICLE-APO/FICHIERDESEQUENCES/HMP_MOCK.V35.FASTA
```

```{r}
# Lire le fichier fasta (si c'est ce que vous voulez faire)
mock.ref_content <- readLines(mock.ref)

# Appliquer la logique avec sapply et grepl
result <- sum(sapply(unqs.mock, function(x) any(grepl(as.character(x), mock.ref_content))))

# Vérifier le résultat
print(result)
```

### Graphiques

```{r}
samples.out <- rownames(seqtab.nochim)
subject <- sapply(strsplit(samples.out, "D"), `[`, 1)
gender <- substr(subject,1,1)
subject <- substr(subject,2,999)
day <- as.integer(sapply(strsplit(samples.out, "D"), `[`, 2))
samdf <- data.frame(Subject=subject, Gender=gender, Day=day)
samdf$When <- "Early"
samdf$When[samdf$Day>100] <- "Late"
rownames(samdf) <- samples.out
```

```{r}
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table(taxa))
ps <- prune_samples(sample_names(ps) != "Mock", ps) # Remove mock sample
```

```{r}
dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))
ps
```

```{r}
plot_richness(ps, x="Day", measures=c("Shannon", "Simpson"), color="When")
```

```{r}
# Transform data to proportions as appropriate for Bray-Curtis distances
ps.prop <- transform_sample_counts(ps, function(otu) otu/sum(otu))
ord.nmds.bray <- ordinate(ps.prop, method="NMDS", distance="bray")
```

```{r}
plot_ordination(ps.prop, ord.nmds.bray, color="When", title="Bray NMDS")
```

```{r}
top20 <- names(sort(taxa_sums(ps), decreasing=TRUE))[1:20]
ps.top20 <- transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
ps.top20 <- prune_taxa(top20, ps.top20)
plot_bar(ps.top20, x="Day", fill="Family") + facet_wrap(~When, scales="free_x")
```

```{r}

```



```{r}
data <- data.frame(
  sample.names
)

# Installer et charger le package cluster pour la distance de Gower
install.packages("cluster")
library(cluster)
```

```{r}
library(cluster)
```

```{r}
data[] <- lapply(data, function(x) if(is.character(x)) factor(x) else x)
# Calculer la matrice de distance de Gower
dist_matrix_gower <- daisy(data, metric = "gower")

# Vérifier la matrice de distance
print(dist_matrix_gower)

# Réaliser la PCoA avec cmdscale
pcoa_result <- cmdscale(dist_matrix_gower, k = 2)  # k = 2 pour obtenir 2 dimensions

# Visualiser les résultats de la PCoA
plot(pcoa_result[, 1], pcoa_result[, 2], 
     xlab = "Dimension 1", ylab = "Dimension 2", 
     main = "PCoA", pch = 19)
```


# Pour les Bactéries

### Définition de la variable et affichage des fichiers

```{r}
paths <- "~/article-apo/Fichierdesequencesbacterie"
list.files(paths)
data2<-read.table(file = "/home/rstudio/article-apo/SraRunTable_bact21.csv", header = TRUE, sep=",") # Création de la table de données
```

### Utiliser une colonne des métadonnées pour définir le nom des échantillons

```{r}
fnFbs <- sort(list.files(paths, pattern="_1.fastq.gz", full.names = TRUE))
fnRbs <- sort(list.files(paths, pattern="_2.fastq.gz", full.names = TRUE))
print(fnFbs)
print(fnRbs)
```

```{r}
sample.names2 <- (data2$Sample.Name)
sample.names2  
```

```{r}
plotQualityProfile(fnFbs[1:2])
```

```{r}
plotQualityProfile(fnRbs[1:20])
```

```{r}
plotQualityProfile(filtFbs[1])
```


### Filtrer et rogner

```{r}
filtFbs <- file.path(paths, "filtered", paste0(sample.names2, "_F_filt.fastq.gz"))
filtRbs <- file.path(paths, "filtered", paste0(sample.names2, "_R_filt.fastq.gz"))
names(filtFbs) <- sample.names2
names(filtRbs) <- sample.names2
```

```{r}
summary(filtFbs)
```

```{r}
out2 <- filterAndTrim(fnFbs, filtFbs, fnRbs, filtRbs, truncLen=c(260,260),
              maxN=0, maxEE=c(4,4), truncQ=4, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) 
head(out2)
```

### Construire un modèle de probabilité d’erreur

```{r}
errF2 <- learnErrors(filtFbs, multithread=TRUE)
```

```{r}
errR2 <- learnErrors(filtRbs, multithread=TRUE)
```

```{r}
plotErrors(errF, nominalQ=TRUE)
```

### Application de l’algorithme d’interférence d’échantillon de base aux données de séqences filtrées et découpées

```{r}
dadaFbs <- dada2(filtFbs, err=errF2, multithread=TRUE)
```

```{r}
dadaRbs <- dada2(filtRbs, err=errR2, multithread=TRUE)
```

```{r}
dadaFs[[1]]
```

### Fusionner les lectures appariées

```{r}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)
```

```{r}
plotQualityProfile(filtFs[1:10])
```

```{r}
plotQualityProfile(filtRs[1:10])
```

```{r}
head(mergers[[1]])
```

### Construire une table de séquences

```{r}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)
```

```{r}
table(nchar(getSequences(seqtab)))
```

### Supprimer les chimères

```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
```

```{r}
dim(seqtab.nochim)
```

```{r}
sum(seqtab.nochim)/sum(seqtab) 
```


